src/topology/builder.rs

```diff
@@ -562,3 +412,242 @@ const fn filter_event_type(event: &Event, data_type: DataType) -> bool {
         DataType::Metric => matches!(event, Event::Metric(_)),
     }
 }
+
+// 128 is an arbitrary, smallish constant
+const TRANSFORM_BATCH_SIZE: usize = 128;
+
+// 64 is an arbitrary, smallish constant
+// TODO: maybe something like 2x num_cpus?
+const CONCURRENCY_LIMIT: usize = 64;
+
+#[derive(Debug, Clone)]
+struct TransformNode {
+    key: ComponentKey,
+    typetag: &'static str,
+    inputs: Vec<OutputId>,
+    input_type: DataType,
+    named_outputs: Vec<String>,
+    enable_concurrency: bool,
+}
+
+fn build_transform(
+    transform: Transform,
+    node: TransformNode,
+    input_rx: BufferStream<Event>,
+) -> (Task, HashMap<OutputId, fanout::ControlChannel>) {
+    match transform {
+        // TODO: avoid the double boxing for function transforms here
+        Transform::Function(t) => build_sync_transform(Box::new(t), node, input_rx),
+        Transform::Synchronous(t) => build_sync_transform(t, node, input_rx),
+        Transform::Task(t) => {
+            build_task_transform(t, input_rx, node.input_type, node.typetag, &node.key)
+        }
+    }
+}
+
+fn build_sync_transform(
+    t: Box<dyn SyncTransform>,
+    node: TransformNode,
+    input_rx: BufferStream<Event>,
+) -> (Task, HashMap<OutputId, fanout::ControlChannel>) {
+    let (outputs, controls) = TransformOutputs::new(node.named_outputs);
+
+    let runner = Runner::new(outputs);
+    let transform = if node.enable_concurrency {
+        runner
+            .run_concurrently(t, input_rx, node.input_type)
+            .boxed()
+    } else {
+        runner.run_inline(t, input_rx, node.input_type).boxed()
+    };
+
+    let mut output_controls = HashMap::new();
+    for (name, control) in controls {
+        match name {
+            None => {
+                output_controls.insert(OutputId::from(&node.key), control);
+            }
+            Some(name) => {
+                output_controls.insert(OutputId::from((&node.key, name)), control);
+            }
+        }
```

Comment:
Can we reduce the common code here?
```suggestion
        let id = name
            .map(|name| OutputId::from((&node.key, name)))
            .unwrap_or_else(|| OutputId::from(&node.key));
        output_controls.insert(id, control);
```